perftest/PerfTest.cpp

```diff
@@ -0,0 +1,275 @@
+        _measurementStarted = true;
+        _startTime = std::chrono::steady_clock::now();
+    } else {
+        abort("Cannot start. Already started.");
+    }
+}
+
+void PerfTest::stopMeasurement()
+{
+    if (!isFinished()) {
+        LOG_DBG("PerfTest Stop measurement");
+        _measurementFinished = true;
+        _stopTime = std::chrono::steady_clock::now();
+        std::chrono::duration elapsedTime = _stopTime - _startTime;
+        double seconds = std::chrono::duration_cast<std::chrono::milliseconds>(elapsedTime).count() / 1000.0;
+        addResult("time", std::to_string(seconds));
+    } else {
+        abort("Cannot stop. Already finished.");
+    }
+}
+
+void PerfTest::connect(const std::string &filePath)
+{
+    LOG_DBG("PerfTest connect " << filePath);
+    _handler->connect(filePath);
+}
+
+void PerfTest::loadDocument(const std::string &filePath)
+{
+    LOG_DBG("PerfTest loadDocument " << filePath);
+    _handler->loadDocument(filePath);
+}
+
+void PerfTest::connectAndLoad(const std::string &filePath)
+{
+    _handler->connect(filePath);
+    _handler->loadDocument(filePath);
+}
+
+void PerfTest::waitForMessage(const std::string &substring, unsigned int timeout)
+{
+    LOG_DBG("PerfTest waitForMessage: " + substring);
+
+    std::chrono::steady_clock::time_point _waitingStart = std::chrono::steady_clock::now();
+    std::chrono::steady_clock::time_point _waitingEnd = _waitingStart + std::chrono::milliseconds(timeout * TRACE_MULTIPLIER);
+
+    bool found;
+    bool timedout;
+    std::chrono::steady_clock::time_point now;
+    do {
+        now = std::chrono::steady_clock::now();
+        found = _handler->hasIncomingMessage(substring);
+        timedout = now > _waitingEnd;
+        if (!found && !timedout) {
+            LOG_TRC("Waiting for message " + substring);
+            // Wait at least 1ms, no longer than 1s
+            // Result: Finds items after 1ms, 2ms, 4ms, 8ms, and so on.
+            // Good balance between fast polling for fast operations (waitForIdle can take <1ms)
+            // and slow polling for slow operations (load document can take >1s)
+            // and everything in between
+            int elapsedms = std::chrono::duration_cast<std::chrono::milliseconds>(now - _waitingStart).count();
+            int sleepms = std::min(std::max(1,elapsedms), 1000);
+            std::this_thread::sleep_for(std::chrono::milliseconds(sleepms));
+        }
+    } while (!found && !timedout);
+
+    double seconds = std::chrono::duration_cast<std::chrono::milliseconds>(now - _waitingStart).count()/1000.0;
+    if (found) {
+        LOG_DBG("PerfTest waitForMessage Found message: " + substring + " after " + std::to_string(seconds) + "s");
+    } else {
+        abort("Timed out waiting for message \"" + substring + "\" after " + std::to_string(seconds) + "s");
+    }
+}
+
+void PerfTest::waitForIdle(size_t timeout)
+{
+    LOG_TRC("PerfTest waitForIdle");
+    sendMessage("uno .uno:WaitForIdle");
```

Comment:
Would it work to not send a specific comment, just make sure nothing arrives on the socket for e.g. 1s?